src/bcrypt_nif_worker.erl

```diff
@@ -14,39 +14,66 @@
          handle_call/3, handle_cast/2, handle_info/2]).
 
 -record(state, {
-          default_log_rounds,
-          context
+          current_ctx_index :: pos_integer(),
+          contexts :: tuple()
          }).
 
-start_link() -> gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+-define(DEFAULT_LOG_ROUNDS, 12).
 
-gen_salt() -> gen_server:call(?MODULE, gen_salt, infinity).
-gen_salt(Rounds) ->
-    gen_server:call(?MODULE, {gen_salt, Rounds}, infinity).
-hashpw(Password, Salt) ->
-    gen_server:call(?MODULE, {hashpw, Password, Salt}, infinity).
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
 
-init([]) ->
-    {ok, Default} = application:get_env(bcrypt, default_log_rounds),
-    Ctx = bcrypt_nif:create_ctx(),
-    {ok, #state{default_log_rounds = Default, context = Ctx}}.
+start_link() ->
+    DefultBCryptQueueNum = erlang:system_info(schedulers),
+    start_link(DefultBCryptQueueNum).
 
-terminate(shutdown, _) -> ok.
+start_link(BCryptQueueNum) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [BCryptQueueNum], []).
 
-handle_call(gen_salt, _From, #state{default_log_rounds = R} = State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
-handle_call({gen_salt, R}, _From, State) ->
-    {reply, {ok, bcrypt_nif:gen_salt(R)}, State};
-handle_call({hashpw, Password, Salt}, _From, #state{context=Ctx}=State) ->
+gen_salt() ->
+    Rounds = application:get_env(bcrypt, default_log_rounds, ?DEFAULT_LOG_ROUNDS),
+    gen_salt(Rounds).
+gen_salt(Rounds) ->
+    {ok, bcrypt_nif:gen_salt(Rounds)}.
+
+hashpw(Password, Salt) ->
     Ref = make_ref(),
-    ok = bcrypt_nif:hashpw(Ctx, Ref, self(), Password, Salt),
+    ok = gen_server:cast(?MODULE, {hashpw, Ref, self(), Password, Salt}),
```

Comment:
Maybe we should also monitor the server, otherwise callers might hang forever if it dies.

Something like:

```erlang
Ref = erlang:monitor(process, whereis(?MODULE), [{alias, reply_demonitor}]),
gen_server:cast(...),
receive
  {Ref, Res} -> Res;
  {'DOWN', Ref, process, _, _} -> exit(noproc)
end.
```

Reply:
💹